proposals/ytt/003-openapi-schemas/README.md

```diff
@@ -0,0 +1,80 @@
+# Export OpenAPI Schema
+
+- Status: **Scoping** | Pre-Alpha | In Alpha | In Beta | GA | Rejected
+- Originating Issue: [ytt#103](https://github.com/vmware-tanzu/carvel-ytt/issues/103)
+
+# Problem Statement
+Configuration Authors want to be able to generate documentation for configuration inputs (for example, Data Values Schema, and Data Values) for their users. Additionally, Configuration Consumers want to be able to validate their configuration inputs by other tools (for example, IDE's and OpenAPI Schema validators). 
+
+# Proposal
+Implement a flag to export an [OpenAPI Schema](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md) yaml document generated from a [Data Values schema](https://carvel.dev/ytt/docs/latest/lang-ref-ytt-schema/) file, and an optional Data Values file. When invoking ytt with configuration inputs and the flag `--schema-inspect-openapi`, the result is the OpenAPI formatted document only. 
```

Comment:
feel free to tell me why this is a (much) worse idea:   what if the flag was more like a `-o openapi` ?

Reply:
-o within ytt controls output format of the data (json, yaml, pos) on top of whats being returned (ie it doesnt control what to return). i dont remember if @cari-lynn and i discussed this variation `ytt -f ... --schema-inspect -o openapi`, but we should give it another thought (perhaps even with -o openapi-short that kills of header or -o openapi for short, and -o openapi-full as long format).

Reply:
ytt's `-o` flag right now supports yaml, json, positions. 

The `-o` flag normally applies to the final result of ytt (like Data Values Schema + Data Values + _templates_). For this feature, we want to stop earlier than the final templated result, and only use the configuration inputs, before they are used in templates. Like inspecting the input values after they are all merged together but before they are used, and we have a command that does something like that now: `--data-values-inspect`. So this flag is like that one. It's also not called out here but I would say its even a possibility that the `--schema-inspect-openapi` and `-o json` could be used together at some point.

Reply:
`-o` flag applies to all output that ytt produces even if we stop within the pipeline midway e.g.

```
$ ytt -v foo=bar --data-values-inspect -ojson
{"foo":"bar"}
```

Reply:
As an example, `-o` flag in kubectl allows `json|yaml|name|go-template|go-template-file|template|templatefile|jsonpath|jsonpath-as-json|jsonpath-file`. These seem to me as _how_ should the data be presented, rather than _what_ data to see. I also think it could get complicated when output may be both of an option to `-o` at the same time. For example `ytt -f . -ojson -o openapi`. I updated the proposal to purpose two flags instead `--schema-inspect-openapi` and `--schema-inspect-openapi-metadata`, but happy to hear more thoughts on `-o`.

Reply:
thanks yall,  I think we can let this thread rest in peace though feel free to reopen if i'm mistaken

Reply:
im still on it...

> These seem to me as how should the data be presented, rather than what data to see.

exactly. openapi is "how" in my mind at this point. it's a representation of ytt schema in a particular format. this is inline with our existing -opos which shows yaml+positions read of the nodes as its own custom format. now does it make sense to present other non-schema data in openapi format, probably not, but that doesnt mean that it shouldnt be a supported format for schema. with that in mind:

```
ytt --files-inspect                 # prints involved files
ytt --data-values-inspect           # prints values files
ytt --data-values-schema-inspect    # prints ytt schema (as ytt schema with annotations)

ytt --data-values-inspect -ojson                # prints values as json
ytt --data-values-schema-inspect -o openapi     # prints data values schema in openapi format
ytt --data-values-schema-inspect -o openapi-full     # prints data values schema in openapi format with headers
```

Reply:
also note that i used `--data-values-schema` instead of `--schema` since we are specifically identifying schema for data values `@data/values-schema` vs some generic schema.

Reply:
> openapi is "how" in my mind at this point

I don't think we want the output of `--data-values-schema-inspect` to be limited by what can be expressed in OpenAPI Schema (e.g. certain kinds of validations).
If that's true, then indicating "OpenAPI" _does_ indicate "what" is being output; it's possibly a subset of what would be output for ytt schema.

It also seems reasonable to me that the user may want the OpenAPI Schema content to be in JSON sometimes and YAML other times. So, these two variables seem naturally independent.

If all that resonates, then there are three different data being output: ytt schema, OpenAPI document, and OpenAPI schema component... and it would then fit in the existing scheme to provide a different `inspect` flag for each:

```
ytt --data-values-schema-inspect                 # prints ytt schema for data values
ytt --data-values-schema-inspect-openapi         # prints OpenAPI document for data values
ytt --data-values-schema-inspect-openapi-schema  # prints OpenAPI schema component for data values
```

_(This is essentially an edited set of what @cari-lynn originally proposed.)_

Reply:
> OpenAPI Schema content to be in JSON sometimes and YAML other times

in my head that would be achieved via stacked output formats e.g. -o openapi,json (or non-cobrafied version like -o openapi -o json). this is pretty common in k8s (-o wide,json)

> I don't think we want the output of --data-values-schema-inspect to be limited by what can be expressed in OpenAPI Schema (e.g. certain kinds of validations).

not sure i understand, by default it would be in native ytt schema format. if you opt into -o openapi, native ytt schema format is represented as openapi. openapi could be seen as an encoding format like base64 or json. in theory we could even decode native ytt schema from openapi format.

Reply:
`-o` is growing on me. While it is true that the OpenAPI Schema will be technically a subset of ytt Schema (since ytt Schema in the future may support validations in Starlark) it seems like those differences are limitation of what the OpenAPI spec can represent, rather than an entirely different definition than a ytt Schema. 

If we think of it like a schema is a way to define default values, structure, and types (and eventually validations and documentation..), then both ytt Schema and OpenAPI Schema do those things, just in a different format. In that case the `-o openapi` flag makes sense to toggle the output format between ytt Schema and OpenAPI.
```
ytt --data-values-schema-inspect -o openapi     # prints data values schema in openapi format
ytt --data-values-schema-inspect -o openapi-full     # prints data values schema in openapi format with headers
```

However, we don't currently have a `--data-values-schema-inspect` flag. What would this flag show without the `-o openapi`? The `--data-values-inspect` flag already shows the configuration inputs merged (including the ytt Schema). Does it make sense to require that `--data-values-schema-inspect` always be used with `-o openapi`?

Reply:
Since the proposal has been updated since this thread, I'm going to resolve it. Feel free to open it up again or start a new one.